base/intfuncs.jl

```diff
@@ -195,29 +195,28 @@ end
 ^(x::Number, p::Integer)  = power_by_squaring(x,p)
 ^(x, p::Integer)          = power_by_squaring(x,p)
 
-# x^p for any literal integer p is lowered to x^Val{p},
+# x^p for any literal integer p is lowered to Base.literal_pow(^, x, Val{p})
 # to enable compile-time optimizations specialized to p.
-# However, we still need a fallback that calls the general ^.
-# To avoid ambiguities for methods that dispatch on the
-# first argument, we dispatch the fallback via internal_pow.
+# However, we still need a fallback that calls the function ^ which may either
+# mean Base.^ or something else, depending on context.
 # We mark these @inline since if the target is marked @inline,
 # we want to make sure that gets propagated,
 # even if it is over the inlining threshold.
-@inline ^(x, p) = internal_pow(x, p)
-@inline internal_pow{p}(x, ::Type{Val{p}}) = x^p
+@inline literal_pow{p}(^, x, ::Type{Val{p}}) = ^(x,p)
```

Comment:
This generic fallback covers many situations, including the case where a user defines a function `^` without importing it from Base. In that case, if we restricted the type of the first argument, there would be no applicable method for `literal_pow`. Keep in mind that the `^` on the right-hand side is really the `^` passed as the first argument to `literal_pow`, not necessarily `Base.:^`.